File: routes/inventory.py
from flask import Blueprint, request, jsonify
from core.db import db
from bson import ObjectId
import json
from bson import json_util
from datetime import datetime

inventory_bp = Blueprint('inventory', __name__)

def parse_json(data):
    return json.loads(json_util.dumps(data))

def calculate_stock(boxes, loose, units_per_box):
    """Standardizes stock to ensure valid box/loose distribution."""
    total_units = (int(boxes) * int(units_per_box)) + int(loose)
    if total_units < 0:
        total_units = 0
    return {
        "boxes_in_stock": total_units // int(units_per_box),
        "loose_units_in_stock": total_units % int(units_per_box),
        "total_absolute_units": total_units
    }

@inventory_bp.route('/api/products', methods=['GET'])
def get_products():
    try:
        products = list(db.products.find({}).sort("updated_at", -1))
        
        total_value = 0
        low_stock_count = 0
        
        for p in products:
            total_units = (p.get('boxes_in_stock', 0) * p.get('units_per_box', 1)) + p.get('loose_units_in_stock', 0)
            total_value += (total_units * p.get('price_per_unit', 0))
            
            # UPDATED: Low stock threshold is now strictly based on units
            # Fallback to old box threshold calculation for backwards compatibility on old records
            threshold = p.get('low_stock_unit_threshold', p.get('low_stock_box_threshold', 1) * p.get('units_per_box', 1))
            
            if total_units <= threshold:
                low_stock_count += 1

        return jsonify({
            "products": parse_json(products),
            "metrics": {
                "total_value": total_value,
                "total_items": len(products),
                "low_stock_count": low_stock_count
            }
        }), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

@inventory_bp.route('/api/products', methods=['POST'])
def create_product():
    data = request.json
    
    if db.products.find_one({"sku": data.get('sku').strip().upper()}):
        return jsonify({"error": f"SKU {data.get('sku')} already exists."}), 400

    stock_data = calculate_stock(
        data.get('boxes_in_stock', 0), 
        data.get('loose_units_in_stock', 0), 
        data.get('units_per_box', 1)
    )

    new_product = {
        "sku": data.get('sku').strip().upper(),
        "name": data.get('name').strip(),
        "category": data.get('category').strip(),
        "price_per_unit": float(data.get('price_per_unit', 0)),
        "boxes_in_stock": stock_data['boxes_in_stock'],
        "units_per_box": int(data.get('units_per_box', 1)),
        "loose_units_in_stock": stock_data['loose_units_in_stock'],
        "low_stock_unit_threshold": int(data.get('low_stock_unit_threshold', 10)),
        "updated_at": datetime.utcnow()
    }
    
    db.products.insert_one(new_product)
    return jsonify({"message": "Product added successfully!"}), 201

@inventory_bp.route('/api/products/<product_id>', methods=['PUT'])
def update_product(product_id):
    data = request.json
    
    existing_sku = db.products.find_one({"sku": data.get('sku').strip().upper(), "_id": {"$ne": ObjectId(product_id)}})
    if existing_sku:
        return jsonify({"error": f"SKU {data.get('sku')} is already used by another product."}), 400

    # Retrieve current product to safely add new stock
    current_product = db.products.find_one({"_id": ObjectId(product_id)})
    if not current_product:
        return jsonify({"error": "Product not found."}), 404

    # Calculate existing absolute total units
    current_units_per_box = current_product.get('units_per_box', 1)
    current_total_units = (current_product.get('boxes_in_stock', 0) * current_units_per_box) + current_product.get('loose_units_in_stock', 0)

    # Calculate the NEW units being added
    new_units_per_box = int(data.get('units_per_box', 1))
    added_total_units = (int(data.get('add_boxes', 0)) * new_units_per_box) + int(data.get('add_loose', 0))

    # Calculate new final stock based on the new box configuration
    final_total_units = current_total_units + added_total_units

    update_fields = {
        "sku": data.get('sku').strip().upper(),
        "name": data.get('name').strip(),
        "category": data.get('category').strip(),
        "price_per_unit": float(data.get('price_per_unit', 0)),
        "boxes_in_stock": final_total_units // new_units_per_box,
        "units_per_box": new_units_per_box,
        "loose_units_in_stock": final_total_units % new_units_per_box,
        "low_stock_unit_threshold": int(data.get('low_stock_unit_threshold', 10)),
        "updated_at": datetime.utcnow()
    }

    db.products.update_one({"_id": ObjectId(product_id)}, {"$set": update_fields})
    return jsonify({"message": "Product updated successfully!"}), 200

@inventory_bp.route('/api/products/<product_id>', methods=['DELETE'])
def delete_product(product_id):
    result = db.products.delete_one({"_id": ObjectId(product_id)})
    if result.deleted_count == 1:
        return jsonify({"message": "Product deleted safely."}), 200
    return jsonify({"error": "Product not found."}), 404

File: static/js/inventory.js
let inventoryData = [];
let showingLowStockOnly = false;

document.addEventListener('DOMContentLoaded', () => {
    loadInventoryData();
});

// --- TOAST NOTIFICATION SYSTEM ---
function showToast(message, type = 'success') {
    const container = document.getElementById('toastContainer');
    const toast = document.createElement('div');
    
    const colors = type === 'error' ? 'bg-red-500' : 'bg-green-500';
    toast.className = `${colors} text-white px-6 py-4 rounded-2xl shadow-2xl font-bold text-sm transform transition-all translate-x-full opacity-0 flex items-center gap-3`;
    toast.innerHTML = `<span class="text-lg">${type === 'error' ? '⚠️' : '✅'}</span> ${message}`;
    
    container.appendChild(toast);
    
    setTimeout(() => {
        toast.classList.remove('translate-x-full', 'opacity-0');
    }, 10);

    setTimeout(() => {
        toast.classList.add('translate-x-full', 'opacity-0');
        setTimeout(() => toast.remove(), 300);
    }, 3000);
}

// --- DATA FETCHING & RENDERING ---
async function loadInventoryData() {
    try {
        const res = await fetch('/api/products');
        if (res.ok) {
            const data = await res.json();
            inventoryData = data.products;
            
            document.getElementById('invTotalValue').innerText = `₹${data.metrics.total_value.toFixed(2)}`;
            document.getElementById('invTotalItems').innerText = data.metrics.total_items;
            document.getElementById('invLowStock').innerText = data.metrics.low_stock_count;

            // NEW: Update Nav Side Panel Alert Badge
            const navBadge = document.getElementById('navLowStockBadge');
            if (navBadge) {
                if (data.metrics.low_stock_count > 0) {
                    navBadge.innerText = data.metrics.low_stock_count;
                    navBadge.classList.remove('hidden');
                } else {
                    navBadge.classList.add('hidden');
                }
            }

            populateCategories();
            filterInventory(); // Triggers sorting and active filters natively
            
            if (typeof loadProducts === "function") loadProducts();
        }
    } catch (err) {
        showToast("Failed to fetch inventory data.", "error");
    }
}

function populateCategories() {
    const filter = document.getElementById('invCategoryFilter');
    const datalist = document.getElementById('categoryList');
    
    const categories = [...new Set(inventoryData.map(p => p.category))];
    
    // Store current value to re-select after populating
    const currentVal = filter.value;
    
    filter.innerHTML = '<option value="all">All Categories</option>';
    categories.forEach(c => {
        filter.innerHTML += `<option value="${c}">${c}</option>`;
    });
    
    // Attempt to restore selection
    if([...filter.options].some(o => o.value === currentVal)) {
        filter.value = currentVal;
    }

    datalist.innerHTML = '';
    categories.forEach(c => {
        datalist.innerHTML += `<option value="${c}">`;
    });
}

function renderInventoryTable(data) {
    const tbody = document.getElementById('inventoryTableBody');
    tbody.innerHTML = '';

    if (data.length === 0) {
        tbody.innerHTML = `<tr><td colspan="6" class="p-8 text-center text-gray-500 font-bold">No products found for this criteria.</td></tr>`;
        return;
    }

    data.forEach(p => {
        const totalUnits = (p.boxes_in_stock * p.units_per_box) + p.loose_units_in_stock;
        
        // Use updated unit threshold logic (with fallback for old data)
        const threshold = p.low_stock_unit_threshold || (p.low_stock_box_threshold * p.units_per_box);
        const isLowStock = totalUnits <= threshold;
        
        const pid = typeof p._id === 'object' ? p._id.$oid : p._id;

        const statusHtml = isLowStock 
            ? `<span class="bg-red-50 text-red-500 border border-red-100 px-3 py-1.5 rounded-xl text-[10px] font-black tracking-wider uppercase shadow-sm">LOW STOCK</span>` 
            : `<span class="bg-green-50 text-green-600 border border-green-100 px-3 py-1.5 rounded-xl text-[10px] font-black tracking-wider uppercase shadow-sm">HEALTHY</span>`;

        const tr = document.createElement('tr');
        tr.className = "hover:bg-white/60 transition-colors group";
        tr.innerHTML = `
            <td class="p-5 font-black text-pink-500">${p.sku}</td>
            <td class="p-5">
                <p class="font-bold text-gray-800">${p.name}</p>
                <p class="text-[10px] text-gray-500 uppercase tracking-widest mt-1">${p.category}</p>
            </td>
            <td class="p-5 font-black text-gray-800">₹${p.price_per_unit.toFixed(2)}</td>
            <td class="p-5">${statusHtml}</td>
            <td class="p-5">
                <p class="text-sm font-bold text-gray-800">${totalUnits} <span class="text-gray-500 font-semibold text-xs">Total Units</span></p>
                <p class="text-[11px] text-gray-500 font-medium mt-1">${p.boxes_in_stock} Box + ${p.loose_units_in_stock} Loose</p>
            </td>
            <td class="p-5 text-right">
                <button onclick='editProduct(${JSON.stringify(p).replace(/'/g, "&#39;")})' class="text-blue-500 hover:text-white font-bold text-xs bg-blue-50 hover:bg-blue-500 px-4 py-2 rounded-xl mr-2 transition-colors border border-blue-100 shadow-sm">Edit</button>
                <button onclick="deleteProduct('${pid}', '${p.name}')" class="text-red-500 hover:text-white font-bold text-xs bg-red-50 hover:bg-red-500 px-4 py-2 rounded-xl transition-colors border border-red-100 shadow-sm">Delete</button>
            </td>
        `;
        tbody.appendChild(tr);
    });
}

// --- FILTER & SIDE PANEL ALERTS ---
function activateLowStockFilter() {
    const invBtn = document.querySelector('[onclick*="inventoryWrapper"]');
    if (invBtn && typeof switchView === 'function') {
        switchView('inventoryWrapper', invBtn);
    }
    showingLowStockOnly = true;
    document.getElementById('clearAlertFilterBtn').classList.remove('hidden');
    filterInventory();
}

function clearLowStockFilter() {
    showingLowStockOnly = false;
    document.getElementById('clearAlertFilterBtn').classList.add('hidden');
    filterInventory();
}

function filterInventory() {
    const query = document.getElementById('invSearch').value.toLowerCase();
    const category = document.getElementById('invCategoryFilter').value;
    const sortBy = document.getElementById('invSortFilter').value;

    let filtered = inventoryData.filter(p => {
        const matchesSearch = p.name.toLowerCase().includes(query) || p.sku.toLowerCase().includes(query);
        const matchesCat = category === 'all' || p.category === category;
        
        let matchesAlerts = true;
        if (showingLowStockOnly) {
            const totalUnits = (p.boxes_in_stock * p.units_per_box) + p.loose_units_in_stock;
            const threshold = p.low_stock_unit_threshold || (p.low_stock_box_threshold * p.units_per_box);
            matchesAlerts = totalUnits <= threshold;
        }

        return matchesSearch && matchesCat && matchesAlerts;
    });

    // NEW: Sorting logic
    if (sortBy === 'price_asc') {
        filtered.sort((a, b) => a.price_per_unit - b.price_per_unit);
    } else if (sortBy === 'price_desc') {
        filtered.sort((a, b) => b.price_per_unit - a.price_per_unit);
    } // 'default' falls back to API order (updated_at)

    renderInventoryTable(filtered);
}

// --- MODAL & FORM LOGIC ---
function openInvModal() {
    document.getElementById('invForm').reset();
    document.getElementById('invProductId').value = '';
    
    // Setup for Adding New Product
    document.getElementById('invModalTitle').innerText = "Add New Product";
    document.getElementById('invSubmitBtn').innerText = "SAVE PRODUCT";
    document.getElementById('invConfigTitle').innerText = "Initial Stock Setup";
    document.getElementById('lblBoxes').innerText = "Initial Boxes";
    document.getElementById('lblLoose').innerText = "Initial Loose Units";
    
    // Hide read-only display
    document.getElementById('invCurrentStockDisplay').classList.add('hidden');
    
    const modal = document.getElementById('invModal');
    const content = document.getElementById('invModalContent');
    modal.classList.remove('hidden');
    setTimeout(() => {
        content.classList.remove('scale-95', 'opacity-0');
        content.classList.add('scale-100', 'opacity-100');
    }, 10);
}

function closeInvModal() {
    const modal = document.getElementById('invModal');
    const content = document.getElementById('invModalContent');
    content.classList.remove('scale-100', 'opacity-100');
    content.classList.add('scale-95', 'opacity-0');
    setTimeout(() => modal.classList.add('hidden'), 200);
}

function editProduct(p) {
    openInvModal();
    const pid = typeof p._id === 'object' ? p._id.$oid : p._id;
    
    document.getElementById('invProductId').value = pid;
    document.getElementById('invSku').value = p.sku;
    document.getElementById('invName').value = p.name;
    document.getElementById('invCategory').value = p.category;
    document.getElementById('invPrice').value = p.price_per_unit;
    document.getElementById('invUnitsPerBox').value = p.units_per_box;
    
    // Migration: use the new unit threshold if available, otherwise calculate equivalent based on old box value
    document.getElementById('invThreshold').value = p.low_stock_unit_threshold || (p.low_stock_box_threshold * p.units_per_box);
    
    // Setup for Editing Existing Product
    document.getElementById('invModalTitle').innerText = "Edit Product & Stock";
    document.getElementById('invSubmitBtn').innerText = "UPDATE PRODUCT";
    document.getElementById('invConfigTitle').innerText = "Add Stock (Leave as 0 to just edit details)";
    document.getElementById('lblBoxes').innerText = "Add Boxes";
    document.getElementById('lblLoose').innerText = "Add Loose Units";
    
    // Default the add inputs to 0
    document.getElementById('invBoxes').value = 0;
    document.getElementById('invLoose').value = 0;

    // Show the Read-Only current stock stats
    const totalUnits = (p.boxes_in_stock * p.units_per_box) + p.loose_units_in_stock;
    document.getElementById('invCurrentStockText').innerText = `${totalUnits} Total Units (${p.boxes_in_stock} Boxes + ${p.loose_units_in_stock} Loose)`;
    document.getElementById('invCurrentStockDisplay').classList.remove('hidden');
}

document.getElementById('invForm').addEventListener('submit', async (e) => {
    e.preventDefault();
    const pid = document.getElementById('invProductId').value;
    const btn = document.getElementById('invSubmitBtn');
    
    const method = pid ? 'PUT' : 'POST';
    const url = pid ? `/api/products/${pid}` : '/api/products';
    
    // Build payload dynamically
    const payload = {
        sku: document.getElementById('invSku').value,
        name: document.getElementById('invName').value,
        category: document.getElementById('invCategory').value,
        price_per_unit: document.getElementById('invPrice').value,
        units_per_box: document.getElementById('invUnitsPerBox').value,
        low_stock_unit_threshold: document.getElementById('invThreshold').value
    };

    if (pid) {
        // We are updating: Send the values as incremental additions
        payload.add_boxes = document.getElementById('invBoxes').value;
        payload.add_loose = document.getElementById('invLoose').value;
    } else {
        // We are creating: Send as initial baseline stock
        payload.boxes_in_stock = document.getElementById('invBoxes').value;
        payload.loose_units_in_stock = document.getElementById('invLoose').value;
    }
    
    btn.disabled = true;
    btn.innerText = "SAVING...";

    try {
        const response = await fetch(url, {
            method: method,
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });
        const data = await response.json();

        if (response.ok) {
            showToast(data.message, 'success');
            closeInvModal();
            loadInventoryData();
        } else {
            showToast(data.error, 'error');
        }
    } catch (err) {
        showToast("Network error occurred.", 'error');
    } finally {
        btn.disabled = false;
        btn.innerText = pid ? "UPDATE PRODUCT" : "SAVE PRODUCT";
    }
});

// --- DELETE LOGIC ---
async function deleteProduct(pid, name) {
    if (!confirm(`⚠️ Are you sure you want to delete "${name}"?\nThis action cannot be undone.`)) return;

    try {
        const response = await fetch(`/api/products/${pid}`, { method: 'DELETE' });
        const data = await response.json();

        if (response.ok) {
            showToast(data.message, 'success');
            loadInventoryData();
        } else {
            showToast(data.error, 'error');
        }
    } catch (err) {
        showToast("Network error while deleting.", 'error');
    }
}